RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitPhase2WeightsProducer.cc

```diff
@@ -0,0 +1,111 @@
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/ConsumesCollector.h"
+
+#include "DataFormats/EcalDigi/interface/EcalDigiCollections.h"
+#include "CondFormats/EcalObjects/interface/EcalCATIAGainRatios.h"
+#include "CondFormats/DataRecord/interface/EcalCATIAGainRatiosRcd.h"
+#include "DataFormats/EcalRecHit/interface/EcalRecHitCollections.h"
+#include "DataFormats/EcalRecHit/interface/EcalUncalibratedRecHit.h"
+#include "DataFormats/EcalDetId/interface/EBDetId.h"
+
+#include <TGraph.h>
+#include <TF1.h>
+
+class EcalUncalibRecHitPhase2WeightsProducer: public edm::stream::EDProducer<>{
+
+
+public:
+  explicit EcalUncalibRecHitPhase2WeightsProducer(const edm::ParameterSet& ps);//, edm::ConsumesCollector& c);
+  void produce(edm::Event& evt, const edm::EventSetup& es) override;
+
+private:
+  
+  float tRise_;
+  float tFall_;  
+  double gainRatios[2]={10,1};  
+  float gratio;
+
+  edm::EDGetTokenT<EBDigiCollectionPh2> ebDigiCollectionToken_;
+  std::string hitCollection_;
+
+};
+
+EcalUncalibRecHitPhase2WeightsProducer::EcalUncalibRecHitPhase2WeightsProducer(const edm::ParameterSet& ps){//, edm::ConsumesCollector& c){
+    
+  hitCollection_ = ps.getParameter<std::string>("EBhitCollection");
+  produces<EBUncalibratedRecHitCollection>(hitCollection_);
+  
+  tRise_ = ps.getParameter<double>("tRise");
+  tFall_ = ps.getParameter<double>("tFall");
+  
+  ebDigiCollectionToken_ = consumes<EBDigiCollectionPh2>(ps.getParameter<edm::InputTag>("BarrelDigis"));
+}
+
+void EcalUncalibRecHitPhase2WeightsProducer::produce(edm::Event& evt, const edm::EventSetup& es){
+  
+  // retrieve digis
+  edm::Handle<EBDigiCollectionPh2> pdigis_;
+  
+  evt.getByToken(ebDigiCollectionToken_, pdigis_);
+  
+  const EBDigiCollectionPh2* pdigis = nullptr;
+  pdigis = pdigis_.product();
+
+  // prepare output
+  auto ebUncalibRechits = std::make_unique<EBUncalibratedRecHitCollection>();
+  
+  for (auto itdg = pdigis->begin(); itdg != pdigis->end(); ++itdg) {
+    
+    EBDataFrame digi(*itdg);
+    EcalDataFrame_Ph2 dataFrame(*itdg);
+    DetId detId(digi.id());
+    
+    bool g1 = false; 
+    std::vector<float> timetrace;
+    std::vector<float> adctrace;
```

Comment:
Since we know the number of samples we could add a `adctrace.reserve(nSamples);` and a `timetrace.reserve(nSamples);`